From: Rajeev Kulkarni <rajeevk@ti.com>
Date: Fri, 21 Jan 2011 09:41:01 +0000 (-0600)
Subject: OMAP4:PM Implement Static dependencies to reduce Power consumption.
X-Git-Url: http://git.omapzoom.org/?p=kernel%2Fomap.git;a=commitdiff_plain;h=dea73678cf13aa7834f757299d892638eb3e88bd;hp=83cb2e808f8bb397ce7a149426639035805034ef

OMAP4:PM Implement Static dependencies to reduce Power consumption.

Create Static dependencies in clockdomains.Since this fuunction
is called only boot time, make it init function.

Change-Id: I369e88db6e8948987a2a80bba19ad77c21b4ae2c
Signed-off-by: Rajeev Kulkarni <rajeevk@ti.com>
---

Index: Kernel/arch/arm/mach-omap2/pm44xx.c
===================================================================
--- Kernel.orig/arch/arm/mach-omap2/pm44xx.c	2011-01-31 18:39:19.389754853 +0530
+++ Kernel/arch/arm/mach-omap2/pm44xx.c	2011-02-02 19:40:28.669754691 +0530
@@ -678,6 +678,55 @@
 }
 
 /**
+ * prcm_clear_statdep_regs :
+ * Clear Static dependencies for the modules
+ * Refer TRM section 3.1.1.1.7.1 for more information
+ */
+static void __init prcm_clear_statdep_regs(void)
+{
+	u32 reg;
+
+	/* MPU towards EMIF, L3_2 and L4CFG clockdomains */
+	reg = OMAP4430_MEMIF_STATDEP_MASK | OMAP4430_L3_2_STATDEP_MASK
+			| OMAP4430_L4CFG_STATDEP_MASK;
+	cm_rmw_mod_reg_bits(reg, 0, OMAP4430_CM1_MPU_MOD,
+		OMAP4_CM_MPU_STATICDEP_OFFSET);
+
+
+	/* Ducati towards EMIF, L3_2, L3_1, L4CFG and L4WKUP clockdomains */
+	reg = OMAP4430_MEMIF_STATDEP_MASK | OMAP4430_L3_1_STATDEP_MASK
+		| OMAP4430_L3_2_STATDEP_MASK | OMAP4430_L4CFG_STATDEP_MASK
+		| OMAP4430_L4WKUP_STATDEP_MASK;
+	cm_rmw_mod_reg_bits(reg, 0, OMAP4430_CM2_CORE_MOD,
+		OMAP4_CM_DUCATI_STATICDEP_OFFSET);
+
+	/* SDMA towards EMIF, L3_2, L3_1, L4CFG, L4WKUP, L3INIT
+	 * and L4PER clockdomains
+	*/
+	reg = OMAP4430_MEMIF_STATDEP_MASK | OMAP4430_L3_1_STATDEP_MASK
+		| OMAP4430_L3_2_STATDEP_MASK | OMAP4430_L4CFG_STATDEP_MASK
+		| OMAP4430_L4WKUP_STATDEP_MASK | OMAP4430_L4PER_STATDEP_MASK
+		| OMAP4430_L3INIT_STATDEP_MASK;
+	cm_rmw_mod_reg_bits(reg, 0, OMAP4430_CM2_CORE_MOD,
+		OMAP4_CM_SDMA_STATICDEP_OFFSET);
+
+	/* C2C towards EMIF clockdomains */
+	cm_rmw_mod_reg_bits(OMAP4430_MEMIF_STATDEP_MASK, 0,
+		OMAP4430_CM2_CORE_MOD, OMAP4_CM_D2D_STATICDEP_OFFSET);
+
+	/* C2C_STATICDEP_RESTORE towards EMIF clockdomains */
+	cm_rmw_mod_reg_bits(OMAP4430_MEMIF_STATDEP_MASK, 0,
+			OMAP4430_CM2_RESTORE_MOD,
+			OMAP4_CM_D2D_STATICDEP_RESTORE_OFFSET);
+
+	 /* SDMA_RESTORE towards EMIF, L3_1, L4_CFG,L4WKUP clockdomains */
+	reg = OMAP4430_MEMIF_STATDEP_MASK | OMAP4430_L3_1_STATDEP_MASK
+		| OMAP4430_L4CFG_STATDEP_MASK | OMAP4430_L4WKUP_STATDEP_MASK;
+	cm_rmw_mod_reg_bits(reg, 0, OMAP4430_CM2_RESTORE_MOD,
+		OMAP4_CM_SDMA_STATICDEP_RESTORE_OFFSET);
+};
+
+/**
  * omap4_pm_init - Init routine for OMAP4 PM
  *
  * Initializes all powerdomain and clockdomain target states
@@ -694,6 +743,7 @@
 
 #ifdef CONFIG_PM
 	prcm_setup_regs();
+	prcm_clear_statdep_regs();
 
 	ret = request_irq(OMAP44XX_IRQ_PRCM,
 			  (irq_handler_t)prcm_interrupt_handler,
